repos/system_upgrade/el8toel9/actors/checkifcfg/libraries/ifcfg.py

```diff
@@ -0,0 +1,145 @@
+            rule_files.append(f.path)
+            continue
+
+        if not f.name.startswith('ifcfg-'):
+            continue
+
+        if f.name == 'ifcfg-lo':
+            continue
+
+        for line in open(f.path):
+            try:
+                (key, value) = line.split('#')[0].strip().split('=')
+            except ValueError:
+                # We're not interested in lines that are not
+                # simple assignments. Play it safe.
+                continue
+
+            if key == 'TYPE':
+                if got_type is None:
+                    got_type = value.lower()
+                elif got_type != value.lower():
+                    bad_type = True
+
+            if key == 'BONDING_MASTER':
+                if got_type is None:
+                    got_type = 'bond'
+                elif got_type != 'bond':
+                    bad_type = True
+
+            if key == 'NM_CONTROLLED' and value.lower() not in TRUE_VALUES:
+                nm_controlled = False
+
+        if got_type not in GOOD_TYPES:
+            bad_type = True
+
+        # Don't bother reporting the file for NM_CONTROLLED=no
+        # if its type is not supportable with NetworkManager anyway
+        if bad_type is True:
+            bad_type_files.append(f.path)
+        elif nm_controlled is False:
+            not_controlled_files.append(f.path)
+
+    if rule_files:
+        title = 'Rule file detected'
+        summary = ('RHEL 9 does not support the legacy network-scripts'
+                   ' package that was deprecated in RHEL 8 in favor of'
+                   ' NetworkManager. Rule files present in the system will'
+                   ' not work with NetworkManager.')
+        remediation = 'Replace rule files with a NetworkManager dispatcher script.'
+        reporting.create_report([
+            reporting.Title(title),
+            reporting.Summary(summary),
+            reporting.Remediation(hint=remediation),
+            reporting.Severity(reporting.Severity.HIGH),
+            reporting.Tags([reporting.Tags.NETWORK, reporting.Tags.SERVICES]),
+            reporting.Flags([reporting.Flags.INHIBITOR]),
+            reporting.RelatedResource('package', 'NetworkManager'),
+            reporting.ExternalLink(
+                title='NetworkManager-dispatcher - Dispatch user scripts for NetworkManager',
+                url='https://networkmanager.dev/docs/api/latest/NetworkManager-dispatcher.html',
+            ),
+        ] + [
+            reporting.RelatedResource('file', fname)
+            for fname in rule_files
+        ])
+
+    if bad_type_files:
+        title = 'Network configuration for unsupported device types detected'
+        summary = ('RHEL 9 does not support the legacy network-scripts'
+                   ' package that was deprecated in RHEL 8 in favor of'
+                   ' NetworkManager. Files for device types that are not'
+                   ' supported by NetworkManager are present in the system.')
```

Comment:
Maybe it is better to include the invalid device types that are present in the system.

Reply:
Well, anything we don't recognize is invalid. The set of valid types is finite, set of invalid ones is potentially infinite.
In other words: with network-scripts someone could just use a TYPE=[pigeons](https://datatracker.ietf.org/doc/html/rfc1149)  and drop in an ifup-pigeons scripts and things would happily work.

Reply:
at this point the plugin knows which unsupported types are present, so the summary could mention those.